<a href="https://colab.research.google.com/github/duduvalle/colabTeste/blob/main/TEEDC_Experimento.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install wisardpkg

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 126 kB 5.1 MB/s 
     |████████████████████████████████| 216 kB 55.8 MB/s 
  ERROR: Failed building wheel for wisardpkg
  Running setup.py clean for wisardpkg
Failed to build wisardpkg
    Running setup.py install for wisardpkg ... done
  DEPRECATION: wisardpkg was installed using the legacy 'setup.py install' method, because a wheel could not be built for it. A possible replacement is to fix the wheel build issue reported above. You can find discussion regarding this at https://github.com/pypa/pip/issues/8368.


In [2]:
import wisardpkg as wp
import csv
import numpy as np
import matplotlib.pyplot as plt
import math
from sklearn.utils import shuffle
import time

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
x_treino = []
y_treino = []

x_teste = []
y_teste = []

with open('dota2Train.csv') as csv_file:
    csv_reader = csv.reader(csv_file, delimiter=',')
    for row in csv_reader:
      x_auxTreino = [0] * 226 
      for i in range(0,117):
        if i == 0:
          y_treino.append(row[i])
        elif i >= 4:
          if row[i] == "1":
            x_auxTreino[i-4] = 1
          elif row[i] == "-1":
            x_auxTreino[i+109] = 1
      x_treino.append(x_auxTreino)

with open('dota2Test.csv') as csv_file:
    csv_reader = csv.reader(csv_file, delimiter=',')
    for row in csv_reader:
      x_auxTeste = [0] * 226 
      for i in range(0,117):
        if i == 0:
          y_teste.append(row[i])
        elif i >= 4:
          if row[i] == "1":
            x_auxTeste[i-4] = 1
          elif row[i] == "-1":
            x_auxTeste[i+109] = 1
      x_teste.append(x_auxTeste)

In [4]:
print(len(x_treino))
print(len(x_teste))
print(x_treino[0])
print(y_treino[0])

92650
[0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
-1


In [5]:
def ksplit(X, y, k=10):
    n = len(y)
    p = math.floor(n/k)

    Xs, ys = [X[p*i: p*(i+1)] for i in range(k)], [y[p*i: p*(i+1)] for i in range(k)] 
    return Xs, ys

x_treinoRandomizado, y_treinoRandomizado = shuffle(x_treino, y_treino)

print(len(x_treinoRandomizado))
print(x_treinoRandomizado[0])
print(y_treinoRandomizado[0])

x_split, y_split = ksplit(x_treinoRandomizado, y_treinoRandomizado)

92650
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
-1


In [6]:
def accuracy(y_pred, y_target):
  return np.where(np.array(y_pred) == np.array(y_target), 1, 0).sum()/len(y_target)

In [ ]:
f = open("drive/MyDrive/1RSP-TEEDC/resultadosBleach.txt", "w")

for i in range(2,65):
  mediaAcuracia = 0
  mediaTempo = 0
  melhorTempo = 0
  melhorAcuracia = 0
  for j in range(10):
    melhorAcuraciaCros = 0
    model = wp.Wisard(i, ignoreZero = False, bleachingActivated=True)
    start_tempo = time.time()

    for k in range(10):
      model.train(x_split[k], y_split[k])
    
    for k in range(10):
      model.leaveMoreOut(x_split[k], y_split[k])
      resultadoAux = accuracy(model.classify(x_teste), y_teste)
      if resultadoAux > melhorAcuraciaCros:
        melhorAcuraciaCros = resultadoAux
      model.train(x_split[k], y_split[k])   

    tempo = time.time() - start_tempo
    resultado = melhorAcuraciaCros
    mediaAcuracia += resultado
    mediaTempo += tempo
    print(tempo)
    if resultado > melhorAcuracia:
      melhorTempo = tempo
      melhorAcuracia = resultado
  mediaTempo = mediaTempo/10
  mediaAcuracia = mediaAcuracia/10
  f.write(f"{mediaAcuracia}, {mediaTempo},\n")
  f.flush()
  print(f"{i} Melhor acuracia: {melhorAcuracia} -- Melhor tempo: {melhorTempo}")
  print(f"{mediaAcuracia} , {mediaTempo}")

205.91531443595886
204.51698970794678
204.56358742713928
209.1177487373352
204.34665369987488
207.47786951065063
212.6715486049652
208.40736198425293
209.86254048347473
207.63609647750854
2 Melhor acuracia: 0.5445890810180688 -- Melhor tempo: 209.1177487373352
0.5371769963085292 , 207.4515711069107
103.31753826141357
98.12598991394043
101.45279717445374
102.51964735984802
98.62742972373962
99.93119835853577
101.84080266952515
100.34929180145264
100.72211623191833
100.27654194831848
3 Melhor acuracia: 0.5292403341752477 -- Melhor tempo: 100.72211623191833
0.524752282883233 , 100.71633534431457
57.355058431625366
55.148531675338745
58.773414850234985
55.93230652809143
58.07547116279602
58.07546377182007
55.737855672836304
59.0479531288147
56.843146562576294
54.589593172073364
4 Melhor acuracia: 0.5204002331455216 -- Melhor tempo: 55.737855672836304
0.5168253351466874 , 56.95787949562073
37.685728788375854
35.41005849838257
35.18945288658142
34.55681872367859
38.273449659347534
34.5899150

2 Melhor acuracia: 0.5478919759082961 -- Melhor tempo: 31.782657384872437 com bleach e ignoraZero

2 Melhor acuracia: 0.5340975325432291 -- Melhor tempo: 4.599377393722534 ignorando zero

12 Melhor acuracia: 0.5383718671070526 -- Melhor tempo: 4.0565876960754395 sem nada

2 Melhor acuracia: 0.5445890810180688 -- Melhor tempo: 209.1177487373352 com bleach

In [8]:
melhorTempo = 0
melhorAcuracia = 0
for i in range(1000):
    model = wp.Wisard(2, ignoreZero = True, bleachingActivated=True)
    start_tempo = time.time()
    model.train(x_treino, y_treino)  
    tempo = time.time() - start_tempo
    resultado = accuracy(model.classify(x_teste), y_teste)
    if resultado > melhorAcuracia:
      melhorTempo = tempo
      melhorAcuracia = resultado
      print(f"{i} acuracia: {melhorAcuracia} com tempo {melhorTempo}")
print(f"Melhor Acuracia: {melhorAcuracia} com tempo {melhorTempo}")

0 acuracia: 0.5383718671070526 com tempo 1.2478654384613037
3 acuracia: 0.5397318826500874 com tempo 1.001648187637329
44 acuracia: 0.5414804740625607 com tempo 1.0009939670562744
62 acuracia: 0.5418690499319992 com tempo 1.0201327800750732
66 acuracia: 0.5425490577035166 com tempo 0.9756152629852295
143 acuracia: 0.5460462405284632 com tempo 1.0244877338409424
190 acuracia: 0.5465319603652613 com tempo 1.1026625633239746
256 acuracia: 0.5493491354186905 com tempo 0.9914247989654541
340 acuracia: 0.5496405673207694 com tempo 1.065824270248413
Melhor Acuracia: 0.5496405673207694 com tempo 1.065824270248413


Melhor Acuracia: 0.5478919759082961 com tempo 1.0303223133087158

Melhor Acuracia: 0.5473091121041384 com tempo 1.0294642448425293

Melhor Acuracia: 0.5475034000388576 com tempo 1.0291740894317627

Melhor Acuracia: 0.5485719836798135 com tempo 1.0291831493377686

Melhor Acuracia: 0.5496405673207694 com tempo 1.065824270248413